admin_guide/out_of_resource_handling.adoc

```diff
@@ -0,0 +1,270 @@
+is consuming more resources than were reserved via `system-reserved` or `kube-reserved` allocations,
+and the node only has `Guaranteed` pod(s) remaining, then the node must choose to evict a
+`Guaranteed` pod in order to preserve node stability, and to limit the impact
+of the unexpected consumption to other `Guaranteed` pod(s).
+
+=== Scheduler
+
+The node will report a condition when a compute resource is under pressure.  The
+scheduler views that condition as a signal to dissuade placing additional
+pods on the node.
+
+.Node Conditions and Scheduler Behavior
+[cols="3a,8a",options="header"]
+|===
+
+|Node Condition |Scheduler Behavior
+
+|`*MemoryPressure*`
+| `BestEffort` pods are not scheduled to the node.
+|===
+
+== Node OOM Behavior
+
+If the node experiences a system OOM (out of memory) event prior to the node is able to reclaim memory,
+the node depends on the [oom_killer](https://lwn.net/Articles/391222/) to respond.
+
+The node sets a `oom_score_adj` value for each container based on the quality of service for the pod.
+
+.Quality of Service OOM Scores
+[cols="3a,8a",options="header"]
+|===
+
+| Quality of Service |oom_score_adj
+
+| `Guaranteed`
+| -998
+
+| `BestEffort`
+| 1000
+
+| `Burstable`
+| min(max(2, 1000 - (1000 * memoryRequestBytes) / machineMemoryCapacityBytes), 999)
+|===
+
+If the node is unable to reclaim memory prior to a node experiencing system OOM, the `oom_killer` will calculate
+an `oom_score` based on the percentage of memory its using on the node, and then add the `oom_score_adj` to get an
+effective `oom_score` for the container, and then kills the container with the highest score.
+
+The intended behavior should be that containers with the lowest quality of service that
+are consuming the largest amount of memory relative to the scheduling request should be killed first in order
+to reclaim memory.
+
+Unlike pod eviction, if a pod container is OOM killed, it may be restarted by the node based on its `RestartPolicy`.
+
+== Best Practices
+
+=== Disable swap memory
+
+Failure to disable swap memory will make the node not recognize it is under *MemoryPressure*.
+
+To take advantage of memory based evictions, operators must disable swap.
+
+For details, see xref:overcommit.adoc#disabling-swap-memory[disabling-swap-memory]
+
+=== Schedulable resources and eviction policies
+
+Let's imagine the following scenario:
+
+* Node memory capacity: `10Gi`
+* Operator wants to reserve 10% of memory capacity for system daemons (kernel, node, etc.)
+* Operator wants to evict pods at 95% memory utilization to reduce thrashing and incidence of system OOM.
+
+To facilitate this scenario, the
+xref:../install_config/master_node_configuration.adoc#install-config-master-node-configuration[node configuration file]
+(the *_node-config.yaml_* file) is modified as follows:
+
+====
+----
+kubeletArguments:
+  eviction-hard:
+    - "memory.available<500Mi"
+  system-reserved:
+    - "1.5Gi"
```

Comment:
Nevermind, I see line 249. Somebody's probably going to mess this up in a cluster...